### ISSUE TITLE
os/types_windows: fileStat.Mode() should use the reparse tag surrogate bit to identify symlinks

### ISSUE THREAD

**Issue by mattico** [golang/go#23684]
**Go Version:** 1.9.3 (affects 1.4+ since cf521ce64f50c4f300294d6b649f34eca87bb8a3)
**Operating System:** Windows 10 x64 build 16299 (Fall Creator's Update)
**What did you do:** attempt to traverse into the OneDrive folder to list files using `filepath.Walk`
**What did you expect to see:** a listing of files inside the OneDrive folder
**What did you see instead:** [traversal stops at the OneDrive folder](https://gist.github.com/mattico/96c5bb7874d0e8ae3d93d5ff7c77d354)

I noticed that the code for [fileStat.Mode() on Windows][filestat-mode] uses the presence of `FILE_ATTRIBUTE_REPARSE_POINT` to determine if a file is a symlink (or rather, acts like a unix symlink). While this is a common method, it is actually incorrect. Since a [reparse point][reparse-point] is simply a hook for a filesystem minifilter driver, their semantics are mostly undefined. When Microsoft added the [Files On-Demand][files-on-demand] feature to OneDrive in the Fall Creator's Update, they used a reparse point which does not behave like a symlink, and thus is not readable using `Readlink()`. This [has caused problems](https://github.com/syncthing/syncthing/issues/4542) for users who attempt to use readlink to traverse directories that report being symlinks.

The good news is that the [surrogate bit](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365197(v=vs.85).aspx) in the reparse tag specifies if "the file or directory represents another named entity in the system". That sounds like a Microsoft definition of symlink-like behavior. I made a program which verified that the surrogate bit is set on reparse points that behave like symlinks:

PATH | REPARSE TAG|IsReparseTagNameSurrogate
--|--|--
D:\symlink|10100000000000000000000000001100|true
D:\symlink_dir|10100000000000000000000000001100|true
D:\junction|10100000000000000000000000000011|true
D:\mountpoint|10100000000000000000000000000011|true
D:\OneDrive|10010000000000000111000000011010|false
D:\hardlink|*hard links do not use reparse points*|*and they don't act like symlinks*

Additionally, [Microsoft has indicated][corefx-issue] that the `IsReparseTagNameSurrogate` macro is the correct way to determine if a reparse point behaves like a symlink.

**TL;DR:**
`filestat.Mode()` should read the reparse tag, and only set `ModeSymlink` if the surrogate bit is set in the reparse tag.

[filestat-mode]: https://github.com/golang/go/blob/6dcaa095c571988f2c1cfc9b914fb5128ab7c5ca/src/os/types_windows.go#L41
[reparse-point]: https://msdn.microsoft.com/en-us/library/windows/desktop/aa365503%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396
[files-on-demand]: https://support.office.com/en-us/article/learn-about-onedrive-files-on-demand-0e6860d3-d9f3-4971-b321-7092438fb38e
[corefx-issue]: https://github.com/dotnet/corefx/issues/24250

**Comment by mattico** [comment]
Moved information to comment on related issue #22579
